proto/buf/alpha/registry/v1alpha1/image.proto

```diff
@@ -29,6 +30,17 @@ message GetImageRequest {
   string owner = 1;
   string repository = 2;
   string reference = 3;
+
+  // Exclude files from imported buf modules in this image.
+  bool exclude_imports = 4;
+
+  // When specified the returned image will only contain the necessary files to
+  // describe these types.
+  repeated string types = 5;
+
+  // When specified, applies as projection mask to the resulting Image, i.e.
+  // only fields specified in this mask are returned.
+  google.protobuf.FieldMask image_mask = 6;
```

Comment:
The only thing that's somewhat concerning about this is that according to fieldmask's documentation:

> ```
> // A repeated field is not allowed except at the last position of a
> // paths string.
> ```

https://github.com/protocolbuffers/protobuf/blob/b696347f80d324c624e65a0b7e3ea5dac7ec5a41/src/google/protobuf/field_mask.proto#L85-L86

Whereas images are repeated ImageFiles at the toplevel, so we wouldn't conform to that. Go doesnt really have a well known fieldmask library so we can handle it however we want, but I'm looking on a bit if this makes sense in any of the other languages.

Reply:
This documentation has confused me in the past in the grpc-gateway, it's surprisingly subtle: https://github.com/grpc-ecosystem/grpc-gateway/issues/1766. I'm still not sure I understand entirely what's up.

Reply:
I dove into fieldmask and its documented restrictions a bit more, here's the summary:

First, highlights from the documented restrictions from https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/field_mask.proto
```
// `FieldMask` represents a set of symbolic field paths, for example:
//
// ...
//
// # Field Masks in Projections
//
// ...
//
// A repeated field is not allowed except at the last position of a
// paths string.
//
// ...
//
// # Field Masks in Update Operations
//
// ...
// 
// If a repeated field is specified for an update operation, new values will
// be appended to the existing repeated field in the target resource. Note that
// a repeated field is only allowed in the last position of a `paths` string.
//
// ...
//
//
// ## Field Mask Verification
//
// The implementation of any API method which has a FieldMask type field in the
// request should verify the included field paths, and return an
// `INVALID_ARGUMENT` error if any path is unmappable.
```

Looking through the fieldmask helper implementations in the protobuf repository, it seems like these are all real limitations in their `Merge` methods:

- c#: https://cs.opensource.google/protobuf/protobuf/+/master:csharp/src/Google.Protobuf/FieldMaskTree.cs;l=273-278;drc=e479410564727d8954e0704254f4345f97e3d844
- c++: https://cs.opensource.google/protobuf/protobuf/+/master:src/google/protobuf/util/field_mask_util.cc;l=468-476;drc=d6699989ab5164cc7778ea4802e49aae51a4f417
- java: https://cs.opensource.google/protobuf/protobuf/+/master:java/util/src/main/java/com/google/protobuf/util/FieldMaskTree.java;l=294-301;drc=de371235c9f50eadad7a3562572b79a9422cd754

Note that the `Merge` method isn't present in go's protobuf implementation. The signature is usually something like `Merge(fullMessage, fieldmask, emptyMessage)`, after which `emptyMessage` will have `fullMessage`'s fields if they were present in field mask.

However, it seems like google itself doesn't really stick to this limitation:

https://google.aip.dev/157

> An API may allow read masks with non-terminal repeated fields (unlike update masks), but is not obligated to do so.

https://google.aip.dev/161

> ```
>   // Reviews for the back cover. The key is the author of the review,
>   // and the value is the text of the review.
>   //
>   // Valid field masks: reviews, reviews.smith, reviews.`John Smith`
>   map<string, string> reviews = 2;
> 
> ```
> 
> > Field masks may permit the use of the * character on a repeated field or map to indicate the specification of particular sub-fields in the collection:
> 
> ```
>   // Valid field masks: authors, authors.*.given_name, authors.*.family_name
>   // Invalid field masks: authors.0, authors.0.given_name
>   repeated Author authors = 2;
> ```

---

I think in short the wkt fieldmask has a:
1. documented limitations on non terminal repeated fields
2. protobuf implementation helpers that conform to 1
3. recommendations that apis accept field masks not conforming to the spec as long as the server implements it.

Point 3 seems like the one we can work with, there's nothing that prevents us from accepting a field mask like `image.file.*.services` and doing the thing we want on the server side. As golang has no fieldmask utils, we're on our own to parse and interpret them. Unfortunately, we would need to prevent the usage of [fieldmaskpb#FieldMask.IsValid](https://pkg.go.dev/google.golang.org/protobuf/types/known/fieldmaskpb#FieldMask.IsValid) as that does only allow repeateds in the last part of the fieldmask: https://github.com/protocolbuffers/protobuf-go/blob/v1.27.1/types/known/fieldmaskpb/field_mask.pb.go#L397-L400.

Still, we could absorb that with our own fieldmask parsing/handling code. Since clients just send field masks, as long as we document what we accept on the server side that could technically work. Unfortunately, languages that have utils for creating fieldmasks like java's `FieldMaskUtil.fromFieldNumbers(Production.class, Production.TITLE_FIELD_NUMBER)` [won't be able to](https://cs.opensource.google/protobuf/protobuf/+/master:java/util/src/main/java/com/google/protobuf/util/FieldMaskUtil.java;l=149-171;drc=3afc8283099dd1747260e088cdf27240a010161e) create the correct `.*.` path parts programatically.

Still, if we accept that on the go side we can make it work. There's a go utility library for fieldmasks that works through repeated fields: https://github.com/mennanov/fmutils/blob/af9ce0f96ea1983d7a3350cb5edebd499f405b0d/fmutils.go#L80-L84. Purely illustrative to what we can do on the server side, we probably don't want to use that for other reasons (it's documented to panic if the mask is invalid). 

If we look at all of the above and decide its acceptable, I would need to look more at if the `.*` (and maybe backtick case) messes with fieldmask has documented special proto json handling.

Overall, this feels like a serious limitation and a mess. I think I'm at the point where a buf fieldmask or a enum with possible view options (similar to described in https://google.aip.dev/157) seems more sane.